File: src/main/scala/2017/day-15.scala
package adventofcode2017

object Day15 {
  val inputA = 289L
  val inputB = 629L
  val factorA = 16807L
  val factorB = 48271L
  val mod = 2147483647

  def next(num: Long, factor: Long, crit: (Long) => Boolean = (num) => true): Long = {
    val value = (num * factor) % mod
    if (crit(value) == true) value else next(value, factor, crit)
  }

  def main(args: Array[String]): Unit = {
    val (_, _, count1) = (0 until 40000000).foldLeft(inputA, inputB, 0) {
      case ((a, b, count), _) => 
        val nextA = next(a, factorA)
        val nextB = next(b, factorB)
        val nextCount = if ((nextA & 0xffff) == (nextB & 0xffff)) count+1 else count
        (nextA, nextB, nextCount)
    }
    println(s"part1 $count1")

    val (_, _, count2) = (0 until 5000000).foldLeft(inputA, inputB, 0) {
      case ((a, b, count), _) => 
        val nextA = next(a, factorA, n => (n % 4) == 0)
        val nextB = next(b, factorB, n => (n % 8) == 0)
        val nextCount = if ((nextA & 0xffff) == (nextB & 0xffff)) count+1 else count
        (nextA, nextB, nextCount)
    }
    println(s"part2 $count2")
  }
}

File: 2017/js/day-15.js
let inputA = 289;
let inputB = 629;
let factorA = 16807;
let factorB = 48271;
const bit32 = 2147483647;
const next = (num, fac, crit) => {
  let val = (num*fac) % bit32;
  while(!crit(val)) {
    val = (val*fac) % bit32;
  }
  return val;
};

let nextA = inputA;
let nextB = inputB;
let count = 0;
for(let i=0; i<40000000; i++) {
  nextA = val = (nextA*factorA) % bit32;
  nextB = val = (nextB*factorB) % bit32;
  if ((nextA & 0xffff) === (nextB & 0xffff)) {
    count++;
  }
}
console.log('part1', count);

nextA = inputA;
nextB = inputB;
let total = 0;
for(let i=0; i<5000000; i++) {
  nextA = next(nextA, factorA, res => res % 4 === 0);
  nextB = next(nextB, factorB, res => res % 8 === 0);
  if ((nextA & 0xffff) === (nextB & 0xffff)) {
    total++;
  }
}
console.log('part2', total);

